### DISCUSSION TITLE
Rootless K3s fails with "operation not permitted" error during mount operations

### DISCUSSION THREAD

**Discussion by nishantmunjal7** [k3s-io/k3s#10851] [Q&A]
When running K3s in rootless mode, I'm encountering an "operation not permitted" error during mount operations. This appears to be related to the handling of mount flags in user namespaces.

**Environmental Info:**
K3s Version: 
k3s version v1.30.4+k3s1 (98262b5d)
go version go1.22.5

Node(s) CPU architecture, OS, and Version: 
Linux ip-**  5.14.0-427.31.1.el9_4.x86_64 #1 SMP PREEMPT_DYNAMIC Fri Aug 9 14:06:03 EDT 2024 x86_64 x86_64 x86_64 GNU/Linux

**Describe the bug:**
failed to create containerd task: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: error mounting \".rancher/k3s/agent/kubelet/pods//volumes/kubernetes.io ~ configmap/config-volume\" to rootfs at \"/etc/coredns\": mount .rancher/k3s/agent/kubelet/pods//volumes/kubernetes.io~configmap/config-volume:/etc/coredns (via /proc/self/fd/6), flags: 0x5021: operation not permitted: unknown" containerID=""

**Steps To Reproduce:**
- 1. Set up K3s in rootless mode
- 2. Create a test directory and mount it with specific options

```
mkdir -p ~/test_dir
sudo mount -o bind,rw,nosuid,nodev ~/somedir /home/user/.rancher
```

- 3. Attempt to start K3s Rootless service from user folder.

**Expected behavior:**
All K3s pods should start and operate normally in rootless mode.

**Actual behavior:**
All pods are in CrashLoopBackOff State

**Additional context / logs:**
I noticed a similar issue for this in [Moby](https://github.com/moby/moby/pull/47558) repo, do we need to implement something similar with K3s containerd version

**Comment by brandond** [comment]
> `sudo mount -o bind,rw,nosuid,nodev ~/somedir /home/user/.rancher`

Why are you bind-mounting the rootless dir with weird options? Why not just set `--data-dir=$HOME/somedir` instead?

**Reply by brandond** [reply]
I'm not convinced this is an actual issue, it is more likely that you're just using unsupported options on the mount point that k3s is running under.

**Reply by nishantmunjal7** [reply]
Thank you for your question. I apologise for the confusion. Let me clarify:

1. The bind mount command I mentioned was not part of my actual setup. It was an attempt to recreate the issue on a working system to better understand the problem.
2. In my real environment, I'm not manually binding any directories. However, my system uses Logical Volume Management (LVM) to separate different parts of the filesystem and mount subdirectories.
3. The original issue occurs in a standard rootless K3s setup, using the default data directory (~/.rancher). I haven't modified this path.
The core issue I'm facing is that K3s, when running in rootless mode, seems to be having trouble with mount operations, possibly related to preserving certain mount flags (nosuid, nodev, noexec).
4. Your suggestion to use `--data-dir` is interesting. Given my LVM setup, i will check and try once what path to use here. Will share if i am able to solve the problem.

Thanks again, appreciate your response!

**Reply by nishantmunjal7** [reply]
Hey @brandond 

The mounting issue seems to be related to an existing open issue in k3s, which you can find [here](https://github.com/k3s-io/k3s/issues/7436)

There's also a related PR in the [opencontainers/runc repo](https://github.com/opencontainers/runc/pull/3805), but the corresponding code doesn't yet exist in k3s's version of runc.

 I have a couple of questions:

1. Is there a plan to merge the upstream runc into k3s anytime soon?
2. I’ve forked the k3s runc repo and added the minimum required code from the original repo to address the issue. Would it be possible for me to submit a PR to k3s's runc repo for the next release?
[Link to my commit that fixes this issue](https://github.com/atlanhq/runc/commit/d034dbb6cea8aa4e6d263883f44497c8969cf91e)

**Reply by brandond** [reply]
> The mounting issue seems to be related to an existing open issue in k3s

As noted at https://github.com/k3s-io/k3s/issues/7436#issuecomment-2063074886 that specific issue has been fixed upstream in v1.31, and will not be backported. Please try with v1.31.

> the corresponding code doesn't yet exist in k3s's version of runc.

Yes it does. Our fork of runc tracks upstream with a single patch applied to allow it to be built on windows. Ref: https://github.com/k3s-io/runc/commits/v1.1.14-k3s1

> Is there a plan to merge the upstream runc into k3s anytime soon?

You're just seeing that our main branch hasn't been synced lately. Check the tags.

> Would it be possible for me to submit a PR to k3s's runc repo for the next release?

Please confirm if this is even necessary, given that it's apparently fixed in Kubernetes v1.31.

**Reply by nishantmunjal7** [reply]
1. It hasn't been fixed with v1.31— I tried it and still encountered the same error
2. Yes, i was aware that runc code is synced with tags. 

I just now realised that the [runc PR for this fix](https://github.com/opencontainers/runc/pull/3805/commits/da780e4d275444e9be5fc75d2005f51d71669a8e) has only been merged into
[v1.2.0-rc.3](https://github.com/opencontainers/runc/releases/tag/v1.2.0-rc.3) 
[v1.2.0-rc.2](https://github.com/opencontainers/runc/releases/tag/v1.2.0-rc.2) 
[v1.2.0-rc.1](https://github.com/opencontainers/runc/releases/tag/v1.2.0-rc.1)

<img width="1413" alt="Screenshot 2024-09-14 at 1 23 49 AM" src="https://github.com/user-attachments/assets/8854e941-9b44-4be3-bc98-0c858e0146cf">


These versions aren’t available in the K3s runc repo yet. 
Do you see any options in this case, or do you suggest maintaining my own fork for now?

**Reply by brandond** [reply]
We're not bundling any of the runc v1.2 release candidates. We won't upgrade until v1.2 has actual releases available, and has been confirmed to work with containerd v1.7. If it will require containerd v2.0 that may be a bit further out.

Does the fix even apply cleanly to runc v1.1 or would it need work to do so?

**Reply by nishantmunjal7** [reply]
Yes, the fix does apply cleanly to runc v1.1. 

I was able to patch a new [tag](https://github.com/atlanhq/runc/tree/v1.1.12-k3s2) - [commit](https://github.com/atlanhq/runc/commit/d034dbb6cea8aa4e6d263883f44497c8969cf91e) and replace the [runc version in K3s](https://github.com/atlanhq/k3s/commit/047721bd207f22a6899a103c483e6123244d7906). After that, all the pods came up fine. They were previously stuck in a CrashLoopBack when I was using K3s v1.31.

**Reply by brandond** [reply]
We're currently shipping v1.1.14 of runc, not v.1.12 - but that looks promising otherwise.

I've reached out to @cyphar to ask about getting that fix backported to v1.1.

**Reply by nishantmunjal7** [reply]
Thanks @brandond. 
I'll hold off for now and will wait for Cyphar's response

**Reply by cyphar** [reply]
That patch is the wrong patch to backport to properly fix the issue, it was later reverted. The correct patch is https://github.com/opencontainers/runc/pull/3967 but it is a bit too large to backport and it had to change some behaviour of mount options in a way that wouldn't be reasonable to backport to a patch release (see the 1.2.0-rc.1 release notes).

We plan to do a 1.2.0 release fairly soon -- if you can verify that 1.2.0-rc.3 works for you that would be very useful.

If it's really necessary, I can try to do a minimal backport of that PR that just adds the fallback code for locked mount options, without fixing the other issues with mount options (which cause the behaviour change). But that code is a little hairy and I'm a bit worried about regressions because the patch would be more of a rewrite than a backport. Also, one of the problems that https://github.com/opencontainers/runc/pull/3967 fixed is that runc would ignore flags that users specified in some circumstances (this is the part of the patch that can't be backported as it could introduce a compatiblity issue) -- if you backport the patch without that part it would make the problem worse (which is why https://github.com/opencontainers/runc/pull/3967 contains a fix for both issues).

**Reply by nishantmunjal7** [reply]
Thanks, @cyphar, for the detailed explanation.

I have verified that 1.2.0-rc.3 works with k3s, and it resolves the issue as expected. In the meantime, I can fork k3s and make it compatible with 1.2.0-rc.3 until the final 1.2.0 release is available.

Appreciate your offer for a minimal backport, but I understand that it could introduce potential regressions. For now, I'll work with a fork for k3s.

**Reply by geraldwuhoo** [reply]
Thanks all for the work on this one. I believe this is blocking my attempt at rootless podman `k3d` as well, since I am seeing the same error signature:

`Error: failed to create containerd task: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: error mounting "/var/lib/kubelet/pods/91f67bdd-aa11-46a4-a00f-ebe6d4744f72/volumes/kubernetes.io~configmap/config-volume" to rootfs at "/etc/coredns": mount /var/lib/kubelet/pods/91f67bdd-aa11-46a4-a00f-ebe6d4744f72/volumes/kubernetes.io~configmap/config-volume:/etc/coredns (via /proc/self/fd/6), flags: 0x5021: operation not permitted: unknown`

I notice that [`runc` 1.2.1 is coming with the next release](https://github.com/k3s-io/k3s/issues/11216), very exciting. I eagerly await the release.
